lnwallet/channel.go

```diff
@@ -2574,32 +2644,59 @@ func createHtlcRetribution(chanState *channeldb.OpenChannel,
 	// HTLC script. Otherwise, is this was an outgoing HTLC that we sent,
 	// then from the PoV of the remote commitment state, they're the
 	// receiver of this HTLC.
-	htlcPkScript, htlcWitnessScript, err := genHtlcScript(
+	scriptInfo, err := genHtlcScript(
 		chanState.ChanType, htlc.Incoming, false,
 		htlc.RefundTimeout, htlc.RHash, keyRing,
 	)
 	if err != nil {
 		return emptyRetribution, err
 	}
 
-	return HtlcRetribution{
-		SignDesc: input.SignDescriptor{
-			KeyDesc: chanState.LocalChanCfg.
-				RevocationBasePoint,
-			DoubleTweak:   commitmentSecret,
-			WitnessScript: htlcWitnessScript,
-			Output: &wire.TxOut{
-				PkScript: htlcPkScript,
-				Value:    int64(htlc.Amt),
-			},
-			HashType: txscript.SigHashAll,
+	signDesc := input.SignDescriptor{
+		KeyDesc: chanState.LocalChanCfg.
+			RevocationBasePoint,
+		DoubleTweak:   commitmentSecret,
+		WitnessScript: scriptInfo.WitnessScriptToSign(),
+		Output: &wire.TxOut{
+			PkScript: scriptInfo.PkScript(),
+			Value:    int64(htlc.Amt),
 		},
+		HashType: txscript.SigHashAll,
+	}
+
+	// For taproot HTLC outputs, we need to set the sign method to key
+	// spend, and also set the tap tweak root needed to derive the proper
+	// private key.
+	if scriptTree, ok := scriptInfo.(input.TapscriptDescriptor); ok {
+		signDesc.SignMethod = input.TaprootKeySpendSignMethod
+
+		signDesc.TapTweak = scriptTree.TapTweak()
+	}
+
+	// The second level script we sign will always be the success path.
+	secondLevelWitnessScript, err := secondLevelScript.WitnessScriptForPath(
+		input.ScriptPathSuccess,
+	)
+	if err != nil {
+		return emptyRetribution, err
+	}
+
+	// If this is a taproot output, we'll also need to obtain the second
+	// level tap tweak as well.
+	var secondLevelTapTweak [32]byte
+	if scriptTree, ok := secondLevelScript.(input.TapscriptDescriptor); ok {
+		copy(secondLevelTapTweak[:], scriptTree.TapTweak())
```

Comment:
should be `secondLevelScript.TapTweak()`

Reply:
No this is a type assertion, the `scriptTree` is bound to that scope, so it's using the interface impl of `secondLevelScript`.